### ISSUE TITLE
Defering shared resource cleanup tasks when using parallel testing

### ISSUE THREAD

**Issue by geekifier** [gruntwork-io/terratest#1106]
I have a testing workflow which functioned as follows.

1. Create an ELB for targeting using a Route53 ELB Alias module.
2. Iterate over test cases using multiple `.tfvars` files, testing different combinations of module's input vars.
3. Deferred cleanup of the ELB resource created in step `1`.

This was easy to accomplish using the `defer` keyword when running tests sequentially.

However, when I started running the tests in step `2` using `Parallel()`, it apparently unblocks the main function to return immediately, causing the ELB to get deleted as soon as it's created.

It is important to note, that my goal here is to create _one_ ELB to be used by _multiple iterations_ of the test cases. I do not want to create multiple ELBs, one for each test run.

How do I handle the deferred cleanup of _global_ resources, which are shared by multiple test invocations?

Relevant code snippets:

```go
func TestR53Example(t *testing.T) {

	// Terraform module's working dir relative to test dir
	tfDir := "../example"

	// tfvars dir relative to TF working dir above
	// ensure there is a trailing slash
	tfVarsDir := "../tests/vars/"

	// Generate a unique suffix to use for test resources
	uniqueId := random.UniqueId()
	aliasName := fmt.Sprintf("terratest-alias-%s", uniqueId)
	elbName := fmt.Sprintf("terratest-elb-%s", uniqueId)

	sess, err := session.NewSession()
	if err != nil {
		fmt.Println("failed to create session,", err)
		return
	}

        // create ELB for our tests to target
	svc := elb.New(sess)

	elbFQDN := createMockResources(t, svc, elbName)

        // at the end of testing, clean up the ELB
	defer cleanupMockResources(t, svc, elbName)

       // iterate over varsFiles and run a module test with each
	varsFiles := getVarsFiles(t, tfVarsDir)

	testId := 1

	for _, varsFile := range varsFiles {

		opts := &terraform.Options{
			TerraformDir: tfDir,
			VarFiles:     []string{tfVarsDir + varsFile.Name()},
			Vars: map[string]interface{}{
				"alias_name":   fmt.Sprintf("%s-%v", aliasName, testId),
				"alias_target": elbFQDN,
			},
		}

		runTests(t, opts)
		testId++

	}
}
```

Inside of runTests()

```go
func runTests(t *testing.T, opts *terraform.Options) {

	t.Run(filepath.Base(opts.VarFiles[0]), func(t *testing.T) {
		t.Parallel()

		tfTempDir := test_structure.CopyTerraformFolderToTemp(t, "../", "./example/")
		// override Terraform dir with temp dir to allow for parallel runs
		opts.TerraformDir = tfTempDir

		logger.Log(t, fmt.Sprintf("Running Terraform with vars files: %s\n\n", opts.VarFiles))

		defer terraform.Destroy(t, opts)

		// You can also use InitAndApply()
		terraform.Init(t, opts)

		// This makes sure there are no changes between 2 Terraform runs
		terraform.ApplyAndIdempotent(t, opts)

               // snipped here for brevity

	})
}
```

**Comment by geekifier** [comment]
This was simpler than expected. Using `t.Cleanup()` outside of the test loop basically takes care of this.
